<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content=CCBR><link rel="shortcut icon" href=../../img/favicon.ico><title>2. Preparing Files - CARLISLE</title><link rel=stylesheet href=../../css/theme.css><link rel=stylesheet href=../../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=../../css/custom.css rel=stylesheet><link href=../../css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "2. Preparing Files";
        var mkdocs_page_input_path = "user-guide/preparing-files.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=../.. class="icon icon-home"> CARLISLE </a><div role=search> <form id=rtd-search-form class=wy-form action=../../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=../..>Home</a> </li> </ul> <p class=caption><span class=caption-text>Usage</span></p> <ul class=current> <li class=toctree-l1><a href=../getting-started/ class="reference internal">1. Getting Started</a> </li> <li class="toctree-l1 current"><a class="reference internal current" href=#>2. Preparing Files</a> <ul class=current> <li class=toctree-l2><a class="reference internal" href=#configs>Configs</a> <ul> <li class=toctree-l3><a class="reference internal" href=#cluster-config>Cluster Config</a> </li> <li class=toctree-l3><a class="reference internal" href=#tools-config>Tools Config</a> </li> <li class=toctree-l3><a class="reference internal" href=#config-yaml>Config YAML</a> <ul> <li class=toctree-l4><a class="reference internal" href=#user-parameters>User Parameters</a> </li> <li class=toctree-l4><a class="reference internal" href=#references>References</a> </li> </ul> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#preparing-manifests>Preparing Manifests</a> <ul> <li class=toctree-l3><a class="reference internal" href=#samples-manifest-required>Samples Manifest (REQUIRED)</a> </li> <li class=toctree-l3><a class="reference internal" href=#contrast-manifest-optional>Contrast Manifest (OPTIONAL)</a> </li> </ul> </li> </ul> </li> <li class=toctree-l1><a href=../run/ class="reference internal">3. Running the Pipeline</a> </li> <li class=toctree-l1><a href=../output/ class="reference internal">4. Expected Output</a> </li> <li class=toctree-l1><a href=../test-info/ class="reference internal">5. Running Test Data</a> </li> </ul> <p class=caption><span class=caption-text>FAQ</span></p> <ul> <li class=toctree-l1><a href=../troubleshooting/ class="reference internal">Troubleshooting</a> </li> <li class=toctree-l1><a href=../contributions/ class="reference internal">Contributions</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=../..>CARLISLE</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=../.. class="icon icon-home" aria-label=Docs></a></li> <li class=breadcrumb-item>Usage</li> <li class="breadcrumb-item active">2. Preparing Files</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/CARLISLE/edit/main/docs/user-guide/preparing-files.md>Edit on CCBR/CARLISLE</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h1 id=preparing-files>Preparing Files<a class=headerlink href=#preparing-files title="Permanent link">&para;</a></h1> <p>The pipeline is controlled through editing configuration and manifest files. Defaults are found in the /WORKDIR/config and /WORKDIR/manifest directories, after initialization.</p> <h2 id=configs>Configs<a class=headerlink href=#configs title="Permanent link">&para;</a></h2> <p>The configuration files control parameters and software of the pipeline. These files are listed below:</p> <ul> <li>config/config.yaml</li> <li>resources/cluster.yaml</li> <li>resources/tools.yaml</li> </ul> <h3 id=cluster-config>Cluster Config<a class=headerlink href=#cluster-config title="Permanent link">&para;</a></h3> <p>The cluster configuration file dictates the resources to be used during submission to Biowulf HPC. There are two different ways to control these parameters - first, to control the default settings, and second, to create or edit individual rules. These parameters should be edited with caution, after significant testing.</p> <h3 id=tools-config>Tools Config<a class=headerlink href=#tools-config title="Permanent link">&para;</a></h3> <p>The tools configuration file dictates the version of each software or program that is being used in the pipeline.</p> <h3 id=config-yaml>Config YAML<a class=headerlink href=#config-yaml title="Permanent link">&para;</a></h3> <p>There are several groups of parameters that are editable for the user to control the various aspects of the pipeline. These are :</p> <ul> <li>Folders and Paths</li> <li>These parameters will include the input and output files of the pipeline, as well as list all manifest names.</li> <li>User parameters</li> <li>These parameters will control the pipeline features. These include thresholds and whether to perform processes.</li> <li>References</li> <li>These parameters will control the location of index files, spike-in references, adaptors and species calling information.</li> </ul> <h4 id=user-parameters>User Parameters<a class=headerlink href=#user-parameters title="Permanent link">&para;</a></h4> <h5 id=spike-in-controls>Spike in Controls<a class=headerlink href=#spike-in-controls title="Permanent link">&para;</a></h5> <p>The pipeline allows for the use of a species specific spike-in control, or the use of normalization via library size. The parameter <code>spikein_genome</code> should be set to the species term used in <code>spikein_reference</code>.</p> <p>For example for ecoli spike-in:</p> <div class=highlight><pre><span></span><code>run_contrasts: true
norm_method: &quot;spikein&quot;
spikein_genome: &quot;ecoli&quot;
spikein_reference:
  ecoli:
    fa: &quot;PIPELINE_HOME/resources/spikein/Ecoli_GCF_000005845.2_ASM584v2_genomic.fna&quot;
</code></pre></div> <p>For example for drosophila spike-in:</p> <div class=highlight><pre><span></span><code>run_contrasts: true
norm_method: &quot;spikein&quot;
spikein_genome: &quot;drosophila&quot;
spikein_reference:
  drosophila:
    fa: &quot;/fdb/igenomes/Drosophila_melanogaster/UCSC/dm6/Sequence/WholeGenomeFasta/genome.fa&quot;
</code></pre></div> <p>If it's determined that the amount of spike-in is not sufficient for the run, a library normaliaztion can be performed.</p> <ol> <li>Complete a CARLISLE run with spike-in set to "Y". This will allow for the complete assessment of the spike-in.</li> <li>Run initial QC analysis on the output data</li> <li>Add the alignment_stats dir to the configuration file.</li> <li>Re-run the CARLISLE pipeline</li> </ol> <h5 id=duplication-status>Duplication Status<a class=headerlink href=#duplication-status title="Permanent link">&para;</a></h5> <p>Users can select duplicated peaks (dedup) or non-deduplicated peaks (no_dedup) through the user parameter.</p> <div class=highlight><pre><span></span><code>dupstatus: &quot;dedup, no_dedup&quot;
</code></pre></div> <h5 id=peak-caller>Peak Caller<a class=headerlink href=#peak-caller title="Permanent link">&para;</a></h5> <p>Three peak callers are available for deployment within the pipeline, with different settings deployed for each caller.</p> <ol> <li><a href=https://github.com/macs3-project/MACS>MACS2</a> is available with two peak calling options: narrowPeak or broadPeak. NOTE: DESeq step generally fails for broadPeak; generally has too many calls.</li> </ol> <div class=highlight><pre><span></span><code>peaktype: &quot;macs2_narrow, macs2_broad,&quot;
</code></pre></div> <ol> <li><a href=https://github.com/FredHutch/SEACR>SEACR</a> is available with four peak calling options: stringent or relaxed parameters, to be paired with "norm" for samples without a spike-in control and "non" for samples with a spikein control</li> </ol> <div class=highlight><pre><span></span><code>peaktype: &quot;seacr_stringent, seacr_relaxed&quot;
</code></pre></div> <ol> <li><a href=https://github.com/maxsonBraunLab/gopeaks>GOPEAKS</a> is available with two peak calling options: narrowpeaks or broadpeaks</li> </ol> <div class=highlight><pre><span></span><code>peaktype: &quot;gopeaks_narrow, gopeaks_broad&quot;
</code></pre></div> <p>A complete list of the available peak calling parameters and the recommended list of parameters is provided below:</p> <table> <thead> <tr> <th>Peak Caller</th> <th>Narrow</th> <th>Broad</th> <th>Normalized, Stringent</th> <th>Normalized, Relaxed</th> <th>Non-Normalized, Stringent</th> <th>Non-Normalized, Relaxed</th> </tr> </thead> <tbody> <tr> <td>Macs2</td> <td>AVAIL</td> <td>AVAIL</td> <td>NA</td> <td>NA</td> <td>NA</td> <td>NA</td> </tr> <tr> <td>SEACR</td> <td>NA</td> <td>NA</td> <td>AVAIL w/o SPIKEIN</td> <td>AVAIL w/o SPIKEIN</td> <td>AVAIL w/ SPIKEIN</td> <td>AVAIL w/ SPIKEIN</td> </tr> <tr> <td>GoPeaks</td> <td>AVAIL</td> <td>AVAIL</td> <td>NA</td> <td>NA</td> <td>NA</td> <td>NA</td> </tr> </tbody> </table> <div class=highlight><pre><span></span><code># Recommended list
### peaktype: &quot;macs2_narrow, macs2_broad, gopeaks_narrow, gopeaks_broad&quot;

# Available list
### peaktype: &quot;macs2_narrow, macs2_broad, seacr_norm_stringent, seacr_norm_relaxed, seacr_non_stringent, seacr_non_relaxed, gopeaks_narrow, gopeaks_broad&quot;
</code></pre></div> <h5 id=macs2-additional-option>Macs2 additional option<a class=headerlink href=#macs2-additional-option title="Permanent link">&para;</a></h5> <p>MACS2 can be run with or without the control. adding a control will increase peak specificity Selecting "Y" for the <code>macs2_control</code> will run the paired control sample provided in the sample manifest</p> <h5 id=quality-tresholds>Quality Tresholds<a class=headerlink href=#quality-tresholds title="Permanent link">&para;</a></h5> <p>Thresholds for quality can be controlled through the <code>quality_thresholds</code> parameter. This must be a list of comma separated values. minimum of numeric value required.</p> <ul> <li>default MACS2 qvalue is 0.05 https://manpages.ubuntu.com/manpages/xenial/man1/macs2_callpeak.1.html</li> <li>default GOPEAKS pvalue is 0.05 https://github.com/maxsonBraunLab/gopeaks/blob/main/README.md</li> <li>default SEACR FDR threshold 1 https://github.com/FredHutch/SEACR/blob/master/README.md</li> </ul> <div class=highlight><pre><span></span><code>#default values
quality_thresholds: &quot;0.1, 0.05, 0.01&quot;
</code></pre></div> <h4 id=references>References<a class=headerlink href=#references title="Permanent link">&para;</a></h4> <p>Additional reference files may be added to the pipeline, if other species were to be used.</p> <p>The absolute file paths which must be included are:</p> <ol> <li>fa: "/path/to/species.fa"</li> <li>blacklist: "/path/to/blacklistbed/species.bed"</li> </ol> <p>The following information must be included:</p> <ol> <li>regions: "list of regions to be included; IE chr1 chr2 chr3"</li> <li>macs2_g: "macs2 genome shorthand; IE mm IE hs"</li> </ol> <h2 id=preparing-manifests>Preparing Manifests<a class=headerlink href=#preparing-manifests title="Permanent link">&para;</a></h2> <p>There are two manifests, one which required for all pipelines and one that is only required if running a differential analysis. These files describe information on the samples and desired contrasts. The paths of these files are defined in the snakemake_config.yaml file. These files are:</p> <ul> <li>samplemanifest</li> <li>contrasts</li> </ul> <h3 id=samples-manifest-required>Samples Manifest (REQUIRED)<a class=headerlink href=#samples-manifest-required title="Permanent link">&para;</a></h3> <p>This manifest will include information to sample level information. It includes the following column headers:</p> <ul> <li>sampleName: the sample name WITHOUT replicate number (IE "SAMPLE")</li> <li>replicateNumber: the sample replicate number (IE "1")</li> <li>isControl: whether the sample should be identified as a control (IE "Y")</li> <li>controlName: the name of the control to use for this sample (IE "CONTROL")</li> <li>controlReplicateNumber: the replicate number of the control to use for this sample (IE "1")</li> <li>path_to_R1: the full path to R1 fastq file (IE "/path/to/sample1.R1.fastq")</li> <li>path_to_R2: the full path to R1 fastq file (IE "/path/to/sample2.R2.fastq")</li> </ul> <p>An example sampleManifest file is shown below:</p> <table> <thead> <tr> <th>sampleName</th> <th>replicateNumber</th> <th>isControl</th> <th>controlName</th> <th>controlReplicateNumber</th> <th>path_to_R1</th> <th>path_to_R2</th> </tr> </thead> <tbody> <tr> <td>53_H3K4me3</td> <td>1</td> <td>N</td> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>PIPELINE_HOME/.test/53_H3K4me3_1.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/53_H3K4me3_1.R2.fastq.gz</td> </tr> <tr> <td>53_H3K4me3</td> <td>2</td> <td>N</td> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>PIPELINE_HOME/.test/53_H3K4me3_2.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/53_H3K4me3_2.R2.fastq.gz</td> </tr> <tr> <td>HN6_H3K4me3</td> <td>1</td> <td>N</td> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>PIPELINE_HOME/.test/HN6_H3K4me3_1.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/HN6_H3K4me3_1.R2.fastq.gz</td> </tr> <tr> <td>HN6_H3K4me3</td> <td>2</td> <td>N</td> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>PIPELINE_HOME/.test/HN6_H3K4me3_2.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/HN6_H3K4me3_2.R2.fastq.gz</td> </tr> <tr> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>Y</td> <td>-</td> <td>-</td> <td>PIPELINE_HOME/.test/HN6_IgG_rabbit_negative_control_1.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/HN6_IgG_rabbit_negative_control_1.R2.fastq.gz</td> </tr> </tbody> </table> <h3 id=contrast-manifest-optional>Contrast Manifest (OPTIONAL)<a class=headerlink href=#contrast-manifest-optional title="Permanent link">&para;</a></h3> <p>This manifest will include sample information to performed differential comparisons.</p> <p>An example contrast file:</p> <table> <thead> <tr> <th>condition1</th> <th>condition2</th> </tr> </thead> <tbody> <tr> <td>MOC1_siSmyd3_2m_25_HCHO</td> <td>MOC1_siNC_2m_25_HCHO</td> </tr> </tbody> </table> <p><strong>Note</strong>: you must have more than one sample per condition in order to perform differential analysis with DESeq2</p> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/CARLISLE class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/CARLISLE</a> </span> <span><a href=../getting-started/ style="color: #fcfcfc">&laquo; Previous</a></span> <span><a href=../run/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=../../js/jquery-3.6.0.min.js></script> <script>var base_url = "../..";</script> <script src=../../js/theme_extra.js></script> <script src=../../js/theme.js></script> <script src=../../js/custom.js></script> <script src=../../search/main.js></script> <script src=../../js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 