from os.path import join
import itertools
import pathlib

def run_pipe_prep(wildcards):
    files=[]
    n=expand(join(RESULTSDIR,"treatment_control_list.txt"))
    files.extend(n)
    return files

def run_library_norm(wildcards):
    files=[]
    if "LIBRARY" in NORM_METHOD:
        lib=expand(join(RESULTSDIR,"alignment_stats","library_scale.tsv"))
        files.extend(lib)
    return files

def run_deeptools_bw(wildcards):
    files=[]
    b1=expand(join(RESULTSDIR,"deeptools","clean","{replicate}.{dupstatus}.clean.bam"),replicate=REPLICATES,dupstatus=DUPSTATUS)
    files.extend(b1)
    b2=expand(join(RESULTSDIR,"deeptools","clean","{replicate}.{dupstatus}.clean.bam.bai"),replicate=REPLICATES,dupstatus=DUPSTATUS)
    files.extend(b2)
    b3=expand(join(RESULTSDIR,"deeptools","clean","{replicate}.{dupstatus}.clean.bigwig"),replicate=REPLICATES,dupstatus=DUPSTATUS)
    files.extend(b3)
    return files

def run_deeptools_heatmap(wildcards):
    files=[]
    GROUP = [a+b for a in TREATMENTS+["all_samples"] for b in ["", ".prot"]]
    p1=expand(join(RESULTSDIR,"deeptools","{group}.{dupstatus}.metagene_heatmap.pdf"),group=GROUP,dupstatus=DUPSTATUS)
    files.extend(p1)
    p2=expand(join(RESULTSDIR,"deeptools","{group}.{dupstatus}.TSS_heatmap.pdf"),group=GROUP,dupstatus=DUPSTATUS)
    files.extend(p2)
    p3=expand(join(RESULTSDIR,"deeptools","{group}.{dupstatus}.metagene_profile.pdf"),group=GROUP,dupstatus=DUPSTATUS)
    files.extend(p3)
    p4=expand(join(RESULTSDIR,"deeptools","{group}.{dupstatus}.TSS_profile.pdf"),group=GROUP,dupstatus=DUPSTATUS)
    files.extend(p4)
    return files

def run_macs2(wildcards):
    files=[]
    if "macs2_narrow" in PEAKTYPE:
        n=expand(join(RESULTSDIR,"peaks","{qthresholds}","macs2","peak_output","{treatment_control_list}.{dupstatus}.narrow.peaks.bed"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_M,dupstatus=DUPSTATUS),
        files.extend(n)
        n2=expand(join(RESULTSDIR,"peaks","{qthresholds}","macs2","peak_output","{treatment_control_list}.{dupstatus}.narrow.peaks.bigbed.gz"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_M,dupstatus=DUPSTATUS),
        files.extend(n2)
        n3=expand(join(RESULTSDIR,"peaks","{qthresholds}","macs2","peak_output","{treatment_control_list}.{dupstatus}.narrow.summits.bed"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_M,dupstatus=DUPSTATUS),
        files.extend(n3)
    if "macs2_broad" in PEAKTYPE:
        b=expand(join(RESULTSDIR,"peaks","{qthresholds}","macs2","peak_output","{treatment_control_list}.{dupstatus}.broad.peaks.bed"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_M,dupstatus=DUPSTATUS),
        files.extend(b)
        b2=expand(join(RESULTSDIR,"peaks","{qthresholds}","macs2","peak_output","{treatment_control_list}.{dupstatus}.broad.peaks.bigbed.gz"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_M,dupstatus=DUPSTATUS),
        files.extend(b2)
    return files

def run_seacr(wildcards):
    files=[]
    if "seacr_stringent" in PEAKTYPE:
        s=expand(join(RESULTSDIR,"peaks","{qthresholds}","seacr","peak_output","{treatment_control_list}.{dupstatus}.stringent.peaks.bed"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS),
        files.extend(s)
        s2=expand(join(RESULTSDIR,"peaks","{qthresholds}","seacr","peak_output","{treatment_control_list}.{dupstatus}.stringent.peaks.bigbed.gz"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS),
        files.extend(s2)
    if "seacr_relaxed" in PEAKTYPE:
        r=expand(join(RESULTSDIR,"peaks","{qthresholds}","seacr","peak_output","{treatment_control_list}.{dupstatus}.relaxed.peaks.bed"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS),
        files.extend(r)
        r2=expand(join(RESULTSDIR,"peaks","{qthresholds}","seacr","peak_output","{treatment_control_list}.{dupstatus}.relaxed.peaks.bigbed.gz"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS),
        files.extend(r2)
    return files

def run_gopeaks(wildcards):
    files=[]
    if "gopeaks_narrow" in PEAKTYPE:
        n=expand(join(RESULTSDIR,"peaks","{qthresholds}","gopeaks","peak_output","{treatment_control_list}.{dupstatus}.narrow.peaks.bed"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS),
        files.extend(n)
        n2=expand(join(RESULTSDIR,"peaks","{qthresholds}","gopeaks","peak_output","{treatment_control_list}.{dupstatus}.narrow.peaks.bigbed.gz"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS),
        files.extend(n2)
    if "gopeaks_broad" in PEAKTYPE:
        b=expand(join(RESULTSDIR,"peaks","{qthresholds}","gopeaks","peak_output","{treatment_control_list}.{dupstatus}.broad.peaks.bed"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS),
        files.extend(b)
        b2=expand(join(RESULTSDIR,"peaks","{qthresholds}","gopeaks","peak_output","{treatment_control_list}.{dupstatus}.broad.peaks.bigbed.gz"),qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS),
        files.extend(b2)
    return files

def run_qc(wildcards):
    files=[]
    if ("gopeaks_narrow" in PEAKTYPE) or ("gopeaks_broad" in PEAKTYPE):
        q=expand(join(RESULTSDIR,'qc', 'multiqc_report_{qthresholds}.html'),qthresholds=QTRESHOLDS)
        files.extend(q)
    else:
        q=join(RESULTSDIR,'qc', 'multiqc_report.html')
        files.append(q)

    if ("SPIKEIN" == NORM_METHOD) and ("ecoli" == spikein_genome):
        html=join(RESULTSDIR,'qc',"spikein_qc_report.html"),
        files.append(html)
    return files

def run_contrasts(wildcards):
    files=[]
    if config["run_contrasts"]:
        files.append(join(RESULTSDIR,"replicate_sample.tsv"))

        # inputs for matrix
        inputs=expand(join(RESULTSDIR,"peaks","{qthresholds}","contrasts","{contrast_list}.{dupstatus}","{contrast_list}.{dupstatus}.{peak_caller_type}.txt"),qthresholds=QTRESHOLDS, contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE)
        files.extend(inputs)

        # counts matrix outputs
        cms=expand(join(RESULTSDIR,"peaks","{qthresholds}","contrasts","{contrast_list}.{dupstatus}","{contrast_list}.{dupstatus}.{peak_caller_type}.countsmatrix.csv"),qthresholds=QTRESHOLDS, contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE)
        files.extend(cms)
        fcm=expand(join(RESULTSDIR,"peaks","{qthresholds}","contrasts","{contrast_list}.{dupstatus}","{contrast_list}.{dupstatus}.{peak_caller_type}.fragmentscountsmatrix.csv"),qthresholds=QTRESHOLDS, contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE)
        files.extend(fcm)
        si=expand(join(RESULTSDIR,"peaks","{qthresholds}","contrasts","{contrast_list}.{dupstatus}","{contrast_list}.{dupstatus}.{peak_caller_type}.sampleinfo.csv"),qthresholds=QTRESHOLDS, contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE)
        files.extend(si)

        # DESEQ outputs
        resultsa=expand(join(RESULTSDIR,"peaks","{qthresholds}","contrasts","{contrast_list}.{dupstatus}","{contrast_list}.{dupstatus}.{peak_caller_type}.AUCbased_diffresults.csv"),qthresholds=QTRESHOLDS, contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE)
        files.extend(resultsa)
        resultsf=expand(join(RESULTSDIR,"peaks","{qthresholds}","contrasts","{contrast_list}.{dupstatus}","{contrast_list}.{dupstatus}.{peak_caller_type}.fragmentsbased_diffresults.csv"),qthresholds=QTRESHOLDS, contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE)
        files.extend(resultsf)

        # diffbb outputs
        abed=expand(join(RESULTSDIR,"peaks","{qthresholds}","contrasts","{contrast_list}.{dupstatus}","{contrast_list}.{dupstatus}.{peak_caller_type}.AUCbased_diffresults.bed"),qthresholds=QTRESHOLDS, contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE)
        files.extend(abed)
        fbed=expand(join(RESULTSDIR,"peaks","{qthresholds}","contrasts","{contrast_list}.{dupstatus}","{contrast_list}.{dupstatus}.{peak_caller_type}.fragmentsbased_diffresults.bed"),qthresholds=QTRESHOLDS, contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE)
        files.extend(fbed)
    return files

def get_motifs(wildcards):
    files=[]
    if ("macs2_narrow" in PEAKTYPE) or ("macs2_broad" in PEAKTYPE):
        anno_m=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","homer","{treatment_control_list}.{dupstatus}.{peak_caller_type}.annotation.txt"),peak_caller="macs2",qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_M,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_M),
        files.extend(anno_m)
        anno_m=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","homer","{treatment_control_list}.{dupstatus}.{peak_caller_type}.motifs","knownResults.html"),peak_caller="macs2",qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_M,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_M),
        files.extend(anno_m)
    if ("gopeaks_narrow" in PEAKTYPE) or ("gopeaks_broad" in PEAKTYPE):
        anno_g=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","homer","{treatment_control_list}.{dupstatus}.{peak_caller_type}.annotation.txt"),peak_caller="gopeaks",qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_G),
        files.extend(anno_g)
        anno_g=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","homer","{treatment_control_list}.{dupstatus}.{peak_caller_type}.motifs","knownResults.html"),peak_caller="gopeaks",qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_G),
        files.extend(anno_g)
    if ("seacr_stringent" in PEAKTYPE) or ("seacr_relaxed" in PEAKTYPE):
        anno_s=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","homer","{treatment_control_list}.{dupstatus}.{peak_caller_type}.annotation.txt"),peak_caller="seacr",qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_S),
        files.extend(anno_s)
        anno_s=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","homer","{treatment_control_list}.{dupstatus}.{peak_caller_type}.motifs","knownResults.html"),peak_caller="seacr",qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_S),
        files.extend(anno_s)
    return files

def get_enrich(wildcards):
    files=[]
    if ("macs2_narrow" in PEAKTYPE) or ("macs2_broad" in PEAKTYPE):
        anno_m=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","homer","enrichment.{dupstatus}.{peak_caller_type}.png"),peak_caller="macs2",qthresholds=QTRESHOLDS,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_M),
        files.extend(anno_m)
    if ("gopeaks_narrow" in PEAKTYPE) or ("gopeaks_broad" in PEAKTYPE):
        anno_g=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","homer","enrichment.{dupstatus}.{peak_caller_type}.png"),peak_caller="gopeaks",qthresholds=QTRESHOLDS,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_G),
        files.extend(anno_g)
    if ("seacr_stringent" in PEAKTYPE) or ("seacr_relaxed" in PEAKTYPE):
        anno_s=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","homer","enrichment.{dupstatus}.{peak_caller_type}.png"),peak_caller="seacr",qthresholds=QTRESHOLDS,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_S),
        files.extend(anno_s)
    return files

def get_combined(wildcards):
    files = []
    if ("macs2_narrow" in PEAKTYPE) or ("macs2_broad" in PEAKTYPE):
        combined_m = expand(join(RESULTSDIR,"peaks","{qthresholds}","macs2","annotation","homer","{treatment_control_list}.{dupstatus}.{peak_caller_type}.annotation_qvalue.tsv"),qthresholds=QTRESHOLDS,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_M,treatment_control_list=TREATMENT_LIST_M)
        files.extend(combined_m)
    return files

def get_rose(wildcards):
    files=[]
    if config['run_rose']:
        if ("macs2_narrow" in PEAKTYPE) or ("macs2_broad" in PEAKTYPE):
            anno_m=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","rose","{treatment_control_list}.{dupstatus}.{peak_caller_type}.{s_dist}","{treatment_control_list}_AllStitched.table.super.summits.bed"),peak_caller="macs2",qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_M,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_M,s_dist=S_DISTANCE),
            files.extend(anno_m)
        if ("gopeaks_narrow" in PEAKTYPE) or ("gopeaks_broad" in PEAKTYPE):
            anno_g=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","rose","{treatment_control_list}.{dupstatus}.{peak_caller_type}.{s_dist}","{treatment_control_list}_AllStitched.table.super.summits.bed"),peak_caller="gopeaks",qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_G,s_dist=S_DISTANCE),
            files.extend(anno_g)
        if ("seacr_stringent" in PEAKTYPE) or ("seacr_relaxed" in PEAKTYPE):
            anno_s=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","rose","{treatment_control_list}.{dupstatus}.{peak_caller_type}.{s_dist}","{treatment_control_list}_AllStitched.table.super.summits.bed"),peak_caller="seacr",qthresholds=QTRESHOLDS,treatment_control_list=TREATMENT_LIST_SG,dupstatus=DUPSTATUS,peak_caller_type=PEAKTYPE_S,s_dist=S_DISTANCE),
            files.extend(anno_s)
    return files

def get_enrichment(wildcards):
    files=[]
    if config["run_contrasts"] and config['run_go_enrichment']:
        if (GENOME == "hg19") or (GENOME == "hg38"):
            if ("macs2_narrow" in PEAKTYPE) or ("macs2_broad" in PEAKTYPE):
                t=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","go_enrichment","{contrast_list}.{dupstatus}.txt"),peak_caller="macs2",qthresholds=QTRESHOLDS,contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS)
                files.extend(t)
                h=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","go_enrichment","{contrast_list}.{dupstatus}.go_enrichment.html"),peak_caller="macs2",qthresholds=QTRESHOLDS,contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS)
                files.extend(h)
            if ("gopeaks_narrow" in PEAKTYPE) or ("gopeaks_broad" in PEAKTYPE):
                t=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","go_enrichment","{contrast_list}.{dupstatus}.txt"),peak_caller="gopeaks",qthresholds=QTRESHOLDS,contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS)
                files.extend(t)
                h=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","go_enrichment","{contrast_list}.{dupstatus}.go_enrichment.html"),peak_caller="gopeaks",qthresholds=QTRESHOLDS,contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS)
                files.extend(h)
            if ("seacr_stringent" in PEAKTYPE) or ("seacr_relaxed" in PEAKTYPE):
                t=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","go_enrichment","{contrast_list}.{dupstatus}.txt"),peak_caller="seacr",qthresholds=QTRESHOLDS,contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS)
                files.extend(t)
                h=expand(join(RESULTSDIR,"peaks","{qthresholds}","{peak_caller}","annotation","go_enrichment","{contrast_list}.{dupstatus}.go_enrichment.html"),peak_caller="seacr",qthresholds=QTRESHOLDS,contrast_list=CONTRAST_LIST,dupstatus=DUPSTATUS)
                files.extend(h)
    return files

include: "rules/init.smk"
include: "rules/align.smk"
include: "rules/peakcalls.smk"
include: "rules/qc.smk"
include: "rules/diff.smk"
include: "rules/annotations.smk"

rule all:
    input:
        ##########################################
        ### required files
        ##########################################
        # manifests
        unpack(run_pipe_prep),

        # ALIGN / deeptools_bw
        unpack(run_deeptools_bw),

        # ALIGN / deeptools_heatmap
        unpack(run_deeptools_heatmap),

        # ALIGN / deeptools_corr
        expand(join(RESULTSDIR,"deeptools","all.{dupstatus}.PearsonCorr.tab"),dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"deeptools","all.{dupstatus}.PCA.tab"),dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"deeptools","all.{dupstatus}.Pearson_heatmap.png"),dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"deeptools","all.{dupstatus}.PearsonPCA.png"),dupstatus=DUPSTATUS),

        # ALIGN / create_library_norm_scales
        unpack(run_library_norm),

        # ALIGN / alignstats
        expand(join(RESULTSDIR,"alignment_stats","{replicate}.alignment_stats.yaml"),replicate=REPLICATES),

        # ALIGN / gather_alignstats
        join(RESULTSDIR,"alignment_stats","alignment_stats.tsv"),

        # PEAKCALLS / macs2_narrow, macs2_broad, seacr_stringent, seacr_relaxed, gopeaks_narrow, gopeaks_broad
        unpack(run_macs2),
        unpack(run_seacr),
        unpack(run_gopeaks),
        join(RESULTSDIR,"peaks","Peak counts.xlsx"),
        join(RESULTSDIR,"peaks","all.peaks.txt"),

        # QC
        unpack(run_qc),

        # DIFF / create_contrast_data_files
        unpack(run_contrasts),

        # ANNOTATION / findMotif, rose, create_contrast_peakcaller_files, go_enrichment
        unpack(get_motifs),
        unpack(get_enrich),
        unpack(get_combined),
        unpack(get_rose),
        unpack(get_enrichment)

on_complete = f"""
for cmd in spooker jobby; do
    if ! command -v $cmd 2>&1 >/dev/null; then
        export PATH="$PATH:{config['ccbr_tools_path']}"
    fi
done
jobby --outerr --tsv logs/snakemake.log | tee logs/snakemake.log.jobby |awk -F"\t" -v OFS="\t" '{{print $2,$3,$18}}' > logs/snakemake.log.jobby.short &&
spooker \
    --outdir {WORKDIR} \
    --name {config['pipeline']} \
    --version {config['version']} \
    --path {config['pipeline_home']} \
    > logs/spooker.log 2>&1

# Check if files were created
if [ -f logs/snakemake.log.jobby ]; then
    echo "✅ logs/snakemake.log.jobby created."
else
    echo "❌ logs/snakemake.log.jobby NOT created."
fi

if [ -f logs/snakemake.log.jobby.short ]; then
    echo "✅ logs/snakemake.log.jobby.short created."
else
    echo "❌ logs/snakemake.log.jobby.short NOT created."
fi

if [ -f logs/spooker.log ]; then
    echo "✅ logs/spooker.log created."
else
    echo "❌ logs/spooker.log NOT created."
fi   
"""

onsuccess:
    print("OnSuccess")
    print(on_complete)
    shell(on_complete)

onerror:
    print("OnError")
    print(on_complete)
    shell(on_complete)

"""
        ##########################################
        ### intermediate files
        ##########################################
        # INIT / create_reference
        join(BOWTIE2_INDEX,"ref.1.bt2"),
        join(BOWTIE2_INDEX,"ref.len"),
        join(BOWTIE2_INDEX,"spikein.len"),

        # ALIGN / trim
        expand(join(RESULTSDIR,"trim","{replicate}.R1.trim.fastq.gz"),replicate=REPLICATES),
        expand(join(RESULTSDIR,"trim","{replicate}.R2.trim.fastq.gz"),replicate=REPLICATES),

        # ALIGN / align
        expand(join(RESULTSDIR,"bam","raw","{replicate}.bam"),replicate=REPLICATES),
        expand(join(RESULTSDIR,"bam","raw","{replicate}.bam.bai"),replicate=REPLICATES),
        expand(join(RESULTSDIR,"bam","raw","{replicate}.bam.flagstat"),replicate=REPLICATES),
        expand(join(RESULTSDIR,"bam","raw","{replicate}.bam.idxstats"),replicate=REPLICATES),

        # ALIGN / filter
        expand(join(RESULTSDIR,"bam","{replicate}.{dupstatus}.bam"),replicate=REPLICATES,dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"bam","{replicate}.{dupstatus}.bam.idxstats"),replicate=REPLICATES,dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"bam","{replicate}.{dupstatus}.bam.flagstat"),replicate=REPLICATES,dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"bam","{replicate}.{dupstatus}.bam.idxstats"),replicate=REPLICATES,dupstatus=DUPSTATUS),

        # ALIGN / alignstats
        join(RESULTSDIR,"alignment_stats","{replicate}.alignment_stats.yaml"),

        # ALIGN / gather_alignstats
        join(RESULTSDIR,"alignment_stats","alignment_stats.tsv"),

        # ALIGN / create_library_norm_scales
        join(RESULTSDIR,"alignment_stats","library_scale.tsv"),

        # ALIGN / bam2bg
        expand(join(RESULTSDIR,"fragments","{replicate}.{dupstatus}.fragments.bed"),replicate=REPLICATES,dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"bedgraph","{replicate}.{dupstatus}.bedgraph"),replicate=REPLICATES,dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"bigwig","{replicate}.{dupstatus}.bigwig"),replicate=REPLICATES,dupstatus=DUPSTATUS),

        # ALIGN / deeptools_heatmap
        expand(join(RESULTSDIR,"deeptools","clean", "{sample}.{dupstatus}.deeptools_prep"), sample=TREATMENTS, dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"deeptools","clean", "{sample}.prot.{dupstatus}.deeptools_prep"), sample=TREATMENTS, dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"deeptools","clean", "all_samples.{dupstatus}.deeptools_prep"), dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"deeptools","clean", "all_samples.prot.{dupstatus}.deeptools_prep"), dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"deeptools","clean", "{group}.{dupstatus}.metagene.mat.gz"),group=[a+b for a in TREATMENTS+["all_samples"] for b in ["", ".prot"]],dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"deeptools","clean", "{group}.{dupstatus}.TSS.mat.gz"),group=[a+b for a in TREATMENTS+["all_samples"] for b in ["", ".prot"]],dupstatus=DUPSTATUS),
        expand(join(RESULTSDIR,"deeptools","clean", "{group}.{dupstatus}.geneinfo.bed"),group=[a+b for a in TREATMENTS+["all_samples"] for b in ["", ".prot"]],dupstatus=DUPSTATUS),
"""
