<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="**C**ut **A**nd **R**un ana**L**ys**IS** pipe**L**in**E**"><meta name=author content=CCBR><link href=../getting-started/ rel=prev><link href=../run/ rel=next><link rel=icon href=../../assets/images/favicon.png><meta name=generator content="mkdocs-1.5.3, mkdocs-material-9.1.6"><title>2. Preparing Files - CARLISLE</title><link rel=stylesheet href=../../assets/stylesheets/main.ded33207.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.a0c5b2b5.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#2-preparing-files class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=CARLISLE class="md-header__button md-logo" aria-label=CARLISLE data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> CARLISLE </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> 2. Preparing Files </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CCBR/CARLISLE title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/CARLISLE </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../getting-started/ class="md-tabs__link md-tabs__link--active"> Usage </a> </li> <li class=md-tabs__item> <a href=../troubleshooting/ class=md-tabs__link> FAQ </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=CARLISLE class="md-nav__button md-logo" aria-label=CARLISLE data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> CARLISLE </label> <div class=md-nav__source> <a href=https://github.com/CCBR/CARLISLE title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/CARLISLE </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> Home </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Usage <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Usage </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../getting-started/ class=md-nav__link> 1. Getting Started </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> 2. Preparing Files <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> 2. Preparing Files </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#21-configs class=md-nav__link> 2.1 Configs </a> <nav class=md-nav aria-label="2.1 Configs"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#211-cluster-config class=md-nav__link> 2.1.1 Cluster Config </a> </li> <li class=md-nav__item> <a href=#212-tools-config class=md-nav__link> 2.1.2 Tools Config </a> </li> <li class=md-nav__item> <a href=#213-config-yaml class=md-nav__link> 2.1.3 Config YAML </a> <nav class=md-nav aria-label="2.1.3 Config YAML"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#2131-user-parameters class=md-nav__link> 2.1.3.1 User Parameters </a> <nav class=md-nav aria-label="2.1.3.1 User Parameters"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#21311-spike-in-controls class=md-nav__link> 2.1.3.1.1 (Spike in Controls) </a> </li> <li class=md-nav__item> <a href=#21312-duplication-status class=md-nav__link> 2.1.3.1.2 Duplication Status </a> </li> <li class=md-nav__item> <a href=#21313-peak-caller class=md-nav__link> 2.1.3.1.3 Peak Caller </a> </li> <li class=md-nav__item> <a href=#213131-macs2-additional-option class=md-nav__link> 2.1.3.1.3.1 Macs2 additional option </a> </li> <li class=md-nav__item> <a href=#21314-quality-tresholds class=md-nav__link> 2.1.3.1.4 Quality Tresholds </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#2132-references class=md-nav__link> 2.1.3.2 References </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#22-preparing-manifests class=md-nav__link> 2.2 Preparing Manifests </a> <nav class=md-nav aria-label="2.2 Preparing Manifests"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#221-samples-manifest-required class=md-nav__link> 2.2.1 Samples Manifest (REQUIRED) </a> </li> <li class=md-nav__item> <a href=#222-contrast-manifest-optional class=md-nav__link> 2.2.2 Contrast Manifest (OPTIONAL) </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../run/ class=md-nav__link> 3. Running the Pipeline </a> </li> <li class=md-nav__item> <a href=../output/ class=md-nav__link> 4. Expected Output </a> </li> <li class=md-nav__item> <a href=../test-info/ class=md-nav__link> 5. Running Test Data </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> FAQ <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../troubleshooting/ class=md-nav__link> Troubleshooting </a> </li> <li class=md-nav__item> <a href=../contributions/ class=md-nav__link> Contributions </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=2-preparing-files>2. Preparing Files<a class=headerlink href=#2-preparing-files title="Permanent link">&para;</a></h1> <p>The pipeline is controlled through editing configuration and manifest files. Defaults are found in the /WORKDIR/config and /WORKDIR/manifest directories, after initialization.</p> <h2 id=21-configs>2.1 Configs<a class=headerlink href=#21-configs title="Permanent link">&para;</a></h2> <p>The configuration files control parameters and software of the pipeline. These files are listed below:</p> <ul> <li>config/config.yaml</li> <li>resources/cluster.yaml</li> <li>resources/tools.yaml</li> </ul> <h3 id=211-cluster-config>2.1.1 Cluster Config<a class=headerlink href=#211-cluster-config title="Permanent link">&para;</a></h3> <p>The cluster configuration file dictates the resouces to be used during submission to Biowulf HPC. There are two differnt ways to control these parameters - first, to control the default settings, and second, to create or edit individual rules. These parameters should be edited with caution, after significant testing.</p> <h3 id=212-tools-config>2.1.2 Tools Config<a class=headerlink href=#212-tools-config title="Permanent link">&para;</a></h3> <p>The tools configuration file dictates the version of each software or program that is being used in the pipeline.</p> <h3 id=213-config-yaml>2.1.3 Config YAML<a class=headerlink href=#213-config-yaml title="Permanent link">&para;</a></h3> <p>There are several groups of parameters that are editable for the user to control the various aspects of the pipeline. These are :</p> <ul> <li>Folders and Paths</li> <li>These parameters will include the input and ouput files of the pipeline, as well as list all manifest names.</li> <li>User parameters</li> <li>These parameters will control the pipeline features. These include thresholds and whether to perform processes.</li> <li>References</li> <li>These parameters will control the location of index files, spike-in references, adaptors and species calling information.</li> </ul> <h4 id=2131-user-parameters>2.1.3.1 User Parameters<a class=headerlink href=#2131-user-parameters title="Permanent link">&para;</a></h4> <h5 id=21311-spike-in-controls>2.1.3.1.1 (Spike in Controls)<a class=headerlink href=#21311-spike-in-controls title="Permanent link">&para;</a></h5> <p>The pipeline allows for the use of a species specific spike-in control, or the use of normalization via library size. The parameter <code>spikein_genome</code> should be set to the species term used in <code>spikein_reference</code>.</p> <p>For example for ecoli spike-in:</p> <div class=highlight><pre><span></span><code>run_contrasts: true
norm_method: &quot;spikein&quot;
spikein_genome: &quot;ecoli&quot;
spikein_reference:
  ecoli:
    fa: &quot;PIPELINE_HOME/resources/spikein/Ecoli_GCF_000005845.2_ASM584v2_genomic.fna&quot;
</code></pre></div> <p>For example for drosophila spike-in:</p> <div class=highlight><pre><span></span><code>run_contrasts: true
norm_method: &quot;spikein&quot;
spikein_genome: &quot;drosophila&quot;
spikein_reference:
  drosophila:
    fa: &quot;/fdb/igenomes/Drosophila_melanogaster/UCSC/dm6/Sequence/WholeGenomeFasta/genome.fa&quot;
</code></pre></div> <p>If it's determined that the amount of spike-in is not sufficient for the run, a library normaliaztion can be performed.</p> <ol> <li>Complete a CARLISLE run with spike-in set to "Y". This will allow for the complete assessment of the spike-in.</li> <li>Run inital QC analysis on the output data</li> <li>Add the alignment_stats dir to the configuration file.</li> <li>Re-run the CARLISLE pipeline</li> </ol> <h5 id=21312-duplication-status>2.1.3.1.2 Duplication Status<a class=headerlink href=#21312-duplication-status title="Permanent link">&para;</a></h5> <p>Users can select duplicated peaks (dedup) or non-deduplicated peaks (no_dedup) through the user parameter.</p> <div class=highlight><pre><span></span><code>dupstatus: &quot;dedup, no_dedup&quot;
</code></pre></div> <h5 id=21313-peak-caller>2.1.3.1.3 Peak Caller<a class=headerlink href=#21313-peak-caller title="Permanent link">&para;</a></h5> <p>Three peak callers are available for deployment within the pipeline, with different settings deployed for each caller.</p> <ol> <li><a href=https://github.com/macs3-project/MACS>MACS2</a> is available with two peak calling options: narrowPeak or broadPeak. NOTE: DESeq step generally fails for broadPeak; generally has too many calls.</li> </ol> <div class=highlight><pre><span></span><code>peaktype: &quot;macs2_narrow, macs2_broad,&quot;
</code></pre></div> <ol> <li><a href=https://github.com/FredHutch/SEACR>SEACR</a> is available with four peak calling options: stringent or relaxed parameters, to be paired with "norm" for samples without a spike-in control and "non" for samples with a spikein control</li> </ol> <div class=highlight><pre><span></span><code>peaktype: &quot;seacr_stringent, seacr_relaxed&quot;
</code></pre></div> <ol> <li><a href=https://github.com/maxsonBraunLab/gopeaks>GOPEAKS</a> is available with two peak calling options: narrowpeaks or broadpeaks</li> </ol> <div class=highlight><pre><span></span><code>peaktype: &quot;gopeaks_narrow, gopeaks_broad&quot;
</code></pre></div> <p>A complete list of the available peak calling parameters and the recommended list of parameters is provided below:</p> <table> <thead> <tr> <th>Peak Caller</th> <th>Narrow</th> <th>Broad</th> <th>Normalized, Stringent</th> <th>Normalized, Relaxed</th> <th>Non-Normalized, Stringent</th> <th>Non-Normalized, Relaxed</th> </tr> </thead> <tbody> <tr> <td>Macs2</td> <td>AVAIL</td> <td>AVAIL</td> <td>NA</td> <td>NA</td> <td>NA</td> <td>NA</td> </tr> <tr> <td>SEACR</td> <td>NA</td> <td>NA</td> <td>AVAIL w/o SPIKEIN</td> <td>AVAIL w/o SPIKEIN</td> <td>AVAIL w/ SPIKEIN</td> <td>AVAIL w/ SPIKEIN</td> </tr> <tr> <td>GoPeaks</td> <td>AVAIL</td> <td>AVAIL</td> <td>NA</td> <td>NA</td> <td>NA</td> <td>NA</td> </tr> </tbody> </table> <div class=highlight><pre><span></span><code># Recommended list
### peaktype: &quot;macs2_narrow, macs2_broad, gopeaks_narrow, gopeaks_broad&quot;

# Available list
### peaktype: &quot;macs2_narrow, macs2_broad, seacr_norm_stringent, seacr_norm_relaxed, seacr_non_stringent, seacr_non_relaxed, gopeaks_narrow, gopeaks_broad&quot;
</code></pre></div> <h5 id=213131-macs2-additional-option>2.1.3.1.3.1 Macs2 additional option<a class=headerlink href=#213131-macs2-additional-option title="Permanent link">&para;</a></h5> <p>MACS2 can be run with or without the control. adding a control will increase peak specificity Selecting "Y" for the <code>macs2_control</code> will run the paired control sample provided in the sample manifest</p> <h5 id=21314-quality-tresholds>2.1.3.1.4 Quality Tresholds<a class=headerlink href=#21314-quality-tresholds title="Permanent link">&para;</a></h5> <p>Thresholds for quality can be controled through the <code>quality_tresholds</code> parameter. This must be a list of comma separated values. minimum of numeric value required.</p> <ul> <li>default MACS2 qvalue is 0.05 <a href=https://manpages.ubuntu.com/manpages/xenial/man1/macs2_callpeak.1.html>https://manpages.ubuntu.com/manpages/xenial/man1/macs2_callpeak.1.html</a></li> <li>default GOPEAKS pvalue is 0.05 <a href=https://github.com/maxsonBraunLab/gopeaks/blob/main/README.md>https://github.com/maxsonBraunLab/gopeaks/blob/main/README.md</a></li> <li>default SEACR FDR threshold 1 <a href=https://github.com/FredHutch/SEACR/blob/master/README.md>https://github.com/FredHutch/SEACR/blob/master/README.md</a></li> </ul> <div class=highlight><pre><span></span><code>#default values
quality_thresholds: &quot;0.1, 0.05, 0.01&quot;
</code></pre></div> <h4 id=2132-references>2.1.3.2 References<a class=headerlink href=#2132-references title="Permanent link">&para;</a></h4> <p>Additional reference files may be added to the pipeline, if other species were to be used.</p> <p>The absolute file paths which must be included are:</p> <ol> <li>fa: "/path/to/species.fa"</li> <li>blacklist: "/path/to/blacklistbed/species.bed"</li> </ol> <p>The following information must be included:</p> <ol> <li>regions: "list of regions to be included; IE chr1 chr2 chr3"</li> <li>macs2_g: "macs2 genome shorthand; IE mm IE hs"</li> </ol> <h2 id=22-preparing-manifests>2.2 Preparing Manifests<a class=headerlink href=#22-preparing-manifests title="Permanent link">&para;</a></h2> <p>There are two manifests, one which required for all pipeliens and one that is only required if running a differential analysis. These files describe information on the samples and desired contrasts. The paths of these files are defined in the snakemake_config.yaml file. These files are:</p> <ul> <li>samplemanifest</li> <li>contrasts</li> </ul> <h3 id=221-samples-manifest-required>2.2.1 Samples Manifest (REQUIRED)<a class=headerlink href=#221-samples-manifest-required title="Permanent link">&para;</a></h3> <p>This manifest will include information to sample level information. It includes the following column headers:</p> <ul> <li>sampleName: the sample name WITHOUT replicate number (IE "SAMPLE")</li> <li>replicateNumber: the sample replicate number (IE "1")</li> <li>isControl: whether the sample should be identified as a control (IE "Y")</li> <li>controlName: the name of the control to use for this sample (IE "CONTROL")</li> <li>controlReplicateNumber: the replicate number of the control to use for this sample (IE "1")</li> <li>path_to_R1: the full path to R1 fastq file (IE "/path/to/sample1.R1.fastq")</li> <li>path_to_R2: the full path to R1 fastq file (IE "/path/to/sample2.R2.fastq")</li> </ul> <p>An example sampleManifest file is shown below:</p> <table> <thead> <tr> <th>sampleName</th> <th>replicateNumber</th> <th>isControl</th> <th>controlName</th> <th>controlReplicateNumber</th> <th>path_to_R1</th> <th>path_to_R2</th> </tr> </thead> <tbody> <tr> <td>53_H3K4me3</td> <td>1</td> <td>N</td> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>PIPELINE_HOME/.test/53_H3K4me3_1.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/53_H3K4me3_1.R2.fastq.gz</td> </tr> <tr> <td>53_H3K4me3</td> <td>2</td> <td>N</td> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>PIPELINE_HOME/.test/53_H3K4me3_2.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/53_H3K4me3_2.R2.fastq.gz</td> </tr> <tr> <td>HN6_H3K4me3</td> <td>1</td> <td>N</td> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>PIPELINE_HOME/.test/HN6_H3K4me3_1.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/HN6_H3K4me3_1.R2.fastq.gz</td> </tr> <tr> <td>HN6_H3K4me3</td> <td>2</td> <td>N</td> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>PIPELINE_HOME/.test/HN6_H3K4me3_2.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/HN6_H3K4me3_2.R2.fastq.gz</td> </tr> <tr> <td>HN6_IgG_rabbit_negative_control</td> <td>1</td> <td>Y</td> <td>-</td> <td>-</td> <td>PIPELINE_HOME/.test/HN6_IgG_rabbit_negative_control_1.R1.fastq.gz</td> <td>PIPELINE_HOME/.test/HN6_IgG_rabbit_negative_control_1.R2.fastq.gz</td> </tr> </tbody> </table> <h3 id=222-contrast-manifest-optional>2.2.2 Contrast Manifest (OPTIONAL)<a class=headerlink href=#222-contrast-manifest-optional title="Permanent link">&para;</a></h3> <p>This manifest will include sample information to performed differential comparisons.</p> <p>An example contrast file:</p> <table> <thead> <tr> <th>condition1</th> <th>condition2</th> </tr> </thead> <tbody> <tr> <td>MOC1_siSmyd3_2m_25_HCHO</td> <td>MOC1_siNC_2m_25_HCHO</td> </tr> </tbody> </table> <p><strong>Note</strong>: you must have more than one sample per condition in order to perform differential analysis with DESeq2</p> <hr> <div class=md-source-file> <small> Last update: 2024-07-16 </small> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2022 CCBR </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=http://bioinformatics.cancer.gov target=_blank rel=noopener title=bioinformatics.cancer.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zm368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zM0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7zM405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.3zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0zm-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.4c73.6 0 133.3 59.7 133.3 133.3 0 14.7-11.9 26.7-26.7 26.7H154.7c-14.7 0-26.7-11.9-26.7-26.7z"/></svg> </a> <a href=https://github.com/CCBR target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/orgs/nciccbr/repositories target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.tabs", "navigation.top", "navigation.indexes", "toc.integrate"], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.51198bba.min.js></script> </body> </html>