---
title: "DifferentialCutAndRun"
output:
  html_document:
    # toc: true
    # toc_float: true
params:
  rawcountsmatrix: "~/CCBR/projects/ccbr1155/CS030586_CARAP/results/peaks/contrasts/siSmyd3_2m_Smyd3_0.25HCHO_500K_vs_siNC_2m_Smyd3_0.25HCHO_500K__no_dedup__norm.relaxed.bed/siSmyd3_2m_Smyd3_0.25HCHO_500K_vs_siNC_2m_Smyd3_0.25HCHO_500K__no_dedup__norm.relaxed.bed_fragmentscountsmatrix.txt"
  coldata: "~/CCBR/projects/ccbr1155/CS030586_CARAP/results/peaks/contrasts/siSmyd3_2m_Smyd3_0.25HCHO_500K_vs_siNC_2m_Smyd3_0.25HCHO_500K__no_dedup__norm.relaxed.bed/siSmyd3_2m_Smyd3_0.25HCHO_500K_vs_siNC_2m_Smyd3_0.25HCHO_500K__no_dedup__norm.relaxed.bed_sampleinfo.txt"
  dupstatus: "no_dedup" # dedup or no_dedup
  condition1: "siSmyd3_2m_Smyd3_0.25HCHO_500K"
  condition2: "siNC_2m_Smyd3_0.25HCHO_500K" # contrasts is condition1 vs condition2 ... pay attention to the order of conditions
  indexcols: "peakID" # comma separated list of indexing columns eg. gene_id,gene_name
  cpm_cutoff: 1
  fdr_cutoff: "0.05"
  log2fc_cutoff: "0.59" # |FC| of 1.5
  results: "~/CCBR/projects/ccbr1155/CS030586_CARAP/results/peaks/contrasts/siSmyd3_2m_Smyd3_0.25HCHO_500K_vs_siNC_2m_Smyd3_0.25HCHO_500K__no_dedup__norm.relaxed.bed/siSmyd3_2m_Smyd3_0.25HCHO_500K_vs_siNC_2m_Smyd3_0.25HCHO_500K__no_dedup__norm.relaxed.bed_fragmentsbased_diffresults.txt.tmp"
  spiked: "Y"
  rawcountsprescaled: "N"
  scalesfbymean: "Y"
  bbpaths: "~/CCBR/projects/ccbr1155/CS030586_CARAP/results/peaks/contrasts/bed_bedgraph_paths.tsv"
  elbowlimits: "~/CCBR/projects/ccbr1155/CS030586_CARAP/results/peaks/contrasts/siSmyd3_2m_Smyd3_0.25HCHO_500K_vs_siNC_2m_Smyd3_0.25HCHO_500K__no_dedup__norm.relaxed.bed/siSmyd3_2m_Smyd3_0.25HCHO_500K_vs_siNC_2m_Smyd3_0.25HCHO_500K__no_dedup__norm.relaxed.bed_fragmentsbased_diffanalysis_elbowlimits.tmp.yaml"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressPackageStartupMessages(library("DESeq2"))
suppressPackageStartupMessages(library("edgeR"))
suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("DT"))
suppressPackageStartupMessages(library("reshape2"))
suppressPackageStartupMessages(library("pander"))
suppressPackageStartupMessages(library("plotly"))
suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("ggfortify"))
suppressPackageStartupMessages(library("ggrepel"))
suppressPackageStartupMessages(library("ELBOW"))
suppressPackageStartupMessages(library("yaml"))
# suppressPackageStartupMessages(library("vsn"))
suppressPackageStartupMessages(library("EnhancedVolcano"))
suppressPackageStartupMessages(library("RUVSeq"))
suppressPackageStartupMessages(library("ChIPseeker"))
suppressPackageStartupMessages(library("TxDb.Mmusculus.UCSC.mm10.knownGene"))
suppressPackageStartupMessages(library("org.Mm.eg.db"))
suppressPackageStartupMessages(library("stringr"))
suppressPackageStartupMessages(library("RColorBrewer"))
# suppressPackageStartupMessages(library("umap"))
```

## Loading SampleInfo and Counts


```{r sampleinfo, include=TRUE, echo=FALSE}
sampleinfo = read.csv(params$coldata,header = TRUE,sep="\t",strip.white = TRUE,check.names = FALSE,colClasses = "character")
# filter based off of params
sampleinfo = sampleinfo[sampleinfo$group == params$condition1 | sampleinfo$group == params$condition2,]
sampleinfo$group = relevel(as.factor(sampleinfo$group),params$condition2)

rawcounts = read.csv(params$rawcountsmatrix,
                     header = TRUE,sep="\t",
                     comment.char = "#", 
                     strip.white = TRUE,
                     check.names = FALSE,
                     colClasses = "character")
rawcounts = as.data.frame(rawcounts)
rawcounts %>% column_to_rownames(var="peakID") -> rawcounts
# filter based off of sampleinfo
rawcounts = rawcounts[,colnames(rawcounts)==sampleinfo$samplename]

# convert character to numeric to integer
x = matrix(as.numeric(as.matrix(rawcounts)),ncol=ncol(rawcounts))
x = matrix(mapply(x,FUN=as.integer),ncol=ncol(rawcounts))
x = as.data.frame(x)
colnames(x) = colnames(rawcounts)
rownames(x) = rownames(rawcounts)
rawcounts = x

sampleinfo=sampleinfo[sampleinfo$samplename==colnames(rawcounts),]
sampleinfo$library_size=colSums(rawcounts)/1e6
sampleinfodf = as.data.frame(sampleinfo)
sampleinfodf$dupstatus = params$dupstatus
rownames(sampleinfo) = sampleinfo$samplename
pander(sampleinfodf,style="rmarkdown")

rawcounts_logcpm = log2(cpm(rawcounts))
cpm_melt=reshape2::melt(rawcounts_logcpm)
colnames(cpm_melt)=c("peakID","samplename","log2cpm")
fdr_cutoff=as.double(params$fdr_cutoff)
log2fc_cutoff=as.double(params$log2fc_cutoff)
```

Total Peaks: `r nrow(rawcounts)` 


Total Samples: `r ncol(rawcounts)`

```{r fdr_check}
print(fdr_cutoff)
print(log2fc_cutoff)
```


```{r cpmplots, echo=FALSE}
ggplot(cpm_melt,aes(x=samplename,y=log2cpm)) + 
  geom_boxplot(fill=as.factor(as.numeric(as.factor(sampleinfo$group))+1)) +
  theme_classic() +
  coord_flip()
  # theme(legend.title = element_blank(),axis.text.x = element_text(angle = 90),legend.text=element_text(size=6),legend.position = "none")
```

## Run DESeq2

```{r dds, include=TRUE, echo=FALSE}
dds <- DESeqDataSetFromMatrix(countData = as.matrix(rawcounts),
                                colData = sampleinfo[,c("samplename","group")],
                                design = ~ group)

if (params$spiked=="Y" & !is.null(params$bbpaths)) {
  bbpaths_df = read.csv(params$bbpaths,
                      header = FALSE,sep="\t",
                      comment.char = "#", 
                      strip.white = TRUE)
  colnames(bbpaths_df)=c("replicate",
                        "sample",
                        "dupstatus",
                        "peaktype",
                        "peakfile",
                        "bedgraph",
                        "scalingfactor")
  # bbpaths_df$cm_col_names=gsub(pattern = ".bedgraph",replacement = "", basename(bbpaths_df$bedgraph))
  sf_df=unique(bbpaths_df[,c("replicate","scalingfactor")])
  dds_cols=colnames(dds)
  sfs=c()
  for (i in dds_cols){
    if (i %in% sf_df$replicate){
      sfs=c(sfs,sf_df[sf_df$replicate==i,"scalingfactor"])
    }
  }
  if (length(sfs)==length(dds_cols)){
# scaling factor magnitudes are variable and depend on the constant used while scaling using spiked-in reads
# DESeq2 size factors are generally hovering around 1
# we try to rescale the scaling factors by dividing them by mean of all scaling factors ... this way they also 
# start hovering around 1 ... based on suggestion from Sohyoung.
    if (params$scalesfbymean == "Y") {
      sfs = sfs/mean(sfs)
    }
# AUC-based counts are prescaled, but fragmentbased counts are not prescaled
    if (params$rawcountsprescaled == "N") {
      rawcounts=round(t(t(rawcounts) * sfs))
      dds <- DESeqDataSetFromMatrix(countData = as.matrix(rawcounts),
                                  colData = sampleinfo[,c("samplename","group")],
                                  design = ~ group)
    }
    DESeq2::sizeFactors(dds)=sfs
  } else {
    print("Samples are spiked, but DESeq2 scaling factors used!!")
  }
}                               
                               
dds <- DESeq(dds)
results <- results(dds)
results_df <- as.data.frame(results)
results_df %>% rownames_to_column(var="peakID") -> results_df
```

### DESeq MAplot

```{r maplot,include=TRUE,echo=FALSE}
DESeq2::plotMA(results)
```

### DESeq PCA

```{r pca,include=TRUE,echo=FALSE}
rld <- vst(dds)
assayrld = as.data.frame(assay(rld))
assayrld$row_variance = rowVars(as.matrix(assayrld))
assayrld = arrange(assayrld,desc(row_variance))
zero_variance_rows=assayrld$row_variance<1e-5
assayrld$row_variance = NULL
assayrld = assayrld[!zero_variance_rows,]
if (nrow(assayrld) > 500){
  assayrld=assayrld[1:500,]
}

pca=prcomp(t(assayrld),scale. = T)
m.pc1 = round(pca$sdev[1]^2/sum(pca$sdev^2)*100,2)
m.pc2 = round(pca$sdev[2]^2/sum(pca$sdev^2)*100,2)
m.pc3 = round(pca$sdev[3]^2/sum(pca$sdev^2)*100,2)
xlab=paste0("PC1(",m.pc1,"%)")
ylab=paste0("PC2(",m.pc2,"%)")
ggplot(pca$x,aes(x=PC1,y=PC2,label=rownames(pca$x)))+geom_point(col=as.factor(as.numeric(as.factor(sampleinfo$group))+1))+
  xlab(xlab)+ylab(ylab)+
  geom_text_repel(max.overlaps = 10,size=2)+
  theme_light()
```



### DESeq Elbow

```{r elbow,include=TRUE,echo=FALSE}
limits=ELBOW::do_elbow_rnaseq(results)
ELBOW::plot_dataset(results, "log2FoldChange", limits$up_limit, limits$low_limit)
write_yaml(limits,file=params$elbowlimits)
```

```{r elbow2,include=TRUE,echo=FALSE}
lim=c(limits$up_limit,limits$low_limit)
lim=as.data.frame(lim)
rownames(lim)=c("UpLimit","DownLimit")
colnames(lim)="log2FC"
lim$FC=2^lim$log2FC
lim["DownLimit","FC"]=-1/lim["DownLimit","FC"]
lim %>% rownames_to_column(var="Limit") -> lim
pander(lim)


# DT::datatable(lim) %>% formatSignif(columns=colnames(lim),digits=4)
```



### DESeq Annotation

```{r annotate,include=TRUE,echo=FALSE}
x = as.data.frame(rownames(results)) 
colnames(x) = c("peakID")
x %>% separate(col = c("peakID"),into = c("chrom","coord"),sep = ":") %>% 
  separate(col = c("coord"),into = c("start","end"),sep = "-") -> x
peaks <- GenomicRanges::makeGRangesFromDataFrame(x)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
peakAnno <- ChIPseeker::annotatePeak(peaks,
                                     tssRegion = c(-2000,200),
                                     TxDb = txdb,
                                     level = "gene",
                                     overlap = "all",
                                     annoDb = "org.Mm.eg.db")
pa <- as.data.frame(peakAnno)
pa$shortAnno=stringr::word(pa$annotation,1)
pa$shortAnno[pa$shortAnno=="5'"]="5'UTR"
pa$shortAnno[pa$shortAnno=="3'"]="3'UTR"
pa$peakID = paste0(pa$seqnames,":",pa$start,"-",pa$end)
results_df = merge(results_df,pa,by=c("peakID"))
write.table(results_df,file=params$results,quote=FALSE,sep="\t",row.names = FALSE,col.names = TRUE)
# fdr_cutoff=0.05
# log2fc_cutoff=0.59
up=as.data.frame(table(results_df[results_df$padj < fdr_cutoff & results_df$log2FoldChange > log2fc_cutoff,]$shortAnno))
down=as.data.frame(table(results_df[results_df$padj < fdr_cutoff & results_df$log2FoldChange < -1*log2fc_cutoff,]$shortAnno))
if(nrow(up)==0){
  up <- data.frame(matrix(ncol = 2, nrow = 0))
}
if(nrow(down)==0){
  down <- data.frame(matrix(ncol = 2, nrow = 0))
}
colnames(up)=c("shortAnno","UP")
colnames(down)=c("shortAnno","DOWN")
deg=as.data.frame(merge(up,down,by=c("shortAnno"),all=TRUE))
deg[is.na(deg)] <- 0
deg %>% column_to_rownames(var="shortAnno") -> deg
deg=cbind(deg,rowSums(deg))
deg=rbind(deg,colSums(deg))
colnames(deg)[length(colnames(deg))]="Total"
rownames(deg)[length(rownames(deg))]="Total"
deg %>% rownames_to_column(var="Annotation") -> deg
pander(deg)
```

### DESeq Volcano

```{r volcano,fig.width=8, fig.height=10,include=TRUE,echo=FALSE}

colors=brewer.pal(7,"Set1")
anno_types=levels(as.factor(results_df$shortAnno))
keyvals=rep("grey",times=nrow(results_df))
names(keyvals)=rep("NS",times=length(keyvals))
for ( i in seq(1,length(anno_types))) {
  keyvals[ abs(results_df$log2FoldChange) > log2fc_cutoff & results_df$padj < fdr_cutoff & results_df$shortAnno == anno_types[i] ] = colors[i]
  names(keyvals)[keyvals == colors[i]] <- anno_types[i]
}
# names(keyvals)[names(keyvals)=="NA"]="NS"
EnhancedVolcano(results_df,
                lab = results_df$SYMBOL,
                x = 'log2FoldChange',
                y = 'padj',
                ylab = bquote(~-Log[10] ~ FDR),
                pCutoff = fdr_cutoff,
                FCcutoff = log2fc_cutoff,
                labSize = 4,
                title = "",
                subtitle = "",
                titleLabSize = 1,
                subtitleLabSize = 1,
                captionLabSize = 10,
                colCustom = keyvals,
                colAlpha = 1,
                # boxedLabels = TRUE,
                # drawConnectors = TRUE,
                legendLabels = c("NS", expression(Log[2] ~ FC), "FDR", expression(FDR ~ and ~ log[2] ~ FC)),
                legendLabSize = 10
  )
```

### DESeq Detailed Results

```{r resultstable,echo=FALSE,include=TRUE}
DT::datatable(results_df)
```